Description: Fix an overflow affecting inodes past 2TiB
Author: Vladimir Serbinenko <phcoder@gmail.com>
Forwarded: yes
Applied-Upstream: http://bazaar.launchpad.net/~vcs-imports/grub/grub2-bzr/revision/3126
Last-Update: 2011-03-29

Index: b/grub-core/fs/ext2.c
===================================================================
--- a/grub-core/fs/ext2.c
+++ b/grub-core/fs/ext2.c
@@ -555,7 +555,7 @@
 
   /* Read the inode.  */
   if (grub_disk_read (data->disk,
-		      ((grub_le_to_cpu32 (blkgrp.inode_table_id) + blkno)
+		      (((grub_disk_addr_t) grub_le_to_cpu32 (blkgrp.inode_table_id) + blkno)
 		        << LOG2_EXT2_BLOCK_SIZE (data)),
 		      EXT2_INODE_SIZE (data) * blkoff,
 		      sizeof (struct grub_ext2_inode), inode))
